core/services/ocr2/plugins/directrequestocr/reporting.go

```diff
@@ -0,0 +1,223 @@
+	if err != nil {
+		r.logger.Error("directRequestReporting Report phase unable to decode query!",
+			commontypes.LogFields{"err": err})
+		return false, nil, err
+	}
+
+	reqIdToResultList := make(map[string][]*ProcessedRequest)
+	for _, id := range queryProto.RequestIDs {
+		reqIdToResultList[string(id)] = []*ProcessedRequest{}
+	}
+
+	for _, ob := range obs {
+		observationProto := &Observation{}
+		err = proto.Unmarshal(ob.Observation, observationProto)
+		if err != nil {
+			r.logger.Error("directRequestReporting Report phase unable to decode observation!",
+				commontypes.LogFields{"err": err, "observer": ob.Observer})
+			continue
+		}
+		for _, res := range observationProto.ProcessedRequests {
+			id := string(res.RequestID)
+			if val, ok := reqIdToResultList[id]; ok {
+				reqIdToResultList[id] = append(val, res)
+			}
+		}
+	}
+
+	// TODO make aggregation modular and configurable with Median as default.
+	// https://app.shortcut.com/chainlinklabs/story/56740/modular-aggregation
+	var aggregated []*ProcessedRequest
+	for _, obsArr := range reqIdToResultList {
+		if len(obsArr) >= 3 {
+			aggregated = append(aggregated, obsArr[0])
+		}
+	}
+
+	r.logger.Debug("directRequestReporting Report phase done", commontypes.LogFields{
+		"nAggregatedRequests": len(aggregated),
+		"reporting":           len(aggregated) > 0,
+	})
+	if len(aggregated) == 0 {
+		return false, nil, nil
+	}
+	reportBytes, err := r.reportCodec.EncodeReport(aggregated)
+	if err != nil {
+		return false, nil, err
+	}
+	return true, reportBytes, nil
+
+}
+
+// ShouldAcceptFinalizedReport() complies with ReportingPlugin
+func (r *directRequestReporting) ShouldAcceptFinalizedReport(ctx context.Context, ts types.ReportTimestamp, report types.Report) (bool, error) {
+	r.logger.Debug("directRequestReporting ShouldAcceptFinalizedReport phase", commontypes.LogFields{
+		"epoch": ts.Epoch,
+		"round": ts.Round,
+	})
+	return true, nil
+}
+
+// ShouldTransmitAcceptedReport() complies with ReportingPlugin
+func (r *directRequestReporting) ShouldTransmitAcceptedReport(ctx context.Context, ts types.ReportTimestamp, report types.Report) (bool, error) {
+	r.logger.Debug("directRequestReporting ShouldTransmitAcceptedReport phase", commontypes.LogFields{
+		"epoch": ts.Epoch,
+		"round": ts.Round,
+	})
+
+	// NOTE: The output of the Report() phase needs to be later decoded by the contract. So unfortunately we
+	// can't use anything more convenient like protobufs but we need to ABI-decode here instead.
+	decoded, err := r.reportCodec.DecodeReport(report)
+	if err != nil {
+		r.logger.Error("unable to decode report built in reporting phase", commontypes.LogFields{"err": err})
+		return false, err
+	}
+
+	allIds := []string{}
+	needTransmissionIds := []string{}
+	for _, item := range decoded {
+		reqIdStr := hex.EncodeToString(item.RequestID)
+		allIds = append(allIds, reqIdStr)
+		prevState, err := r.pluginORM.SetState(sliceToByte32(item.RequestID), directrequestocr.TRANSMITTED)
```

Comment:
Looks like `SetState` should implement a FSM, so all the logic about the proper state transitioning was in one place.

Reply:
Yes, that's the plan. I re-worded the comment below to make it clear. It is a part of the DB/ORM story.